### ISSUE TITLE
Error messages "No error" followed by error detail

### ISSUE THREAD

**Issue by apnadkarni** [bohagan1/TclTLS#1]
Tested on Windows with both VS2022 and Mingw64 both with openssl 3.1.2. SSL_CERT_FILE env set to pem downloaded from curl.se.
 
The following kind of error shows up consistently across multiple tests:

```
==== BadSSL-1.1 1000-sans FAILED
==== Contents of test case:

        badssl 1000-sans.badssl.com

---- Result was:
handshake failed: No error due to "certificate has expired"
---- Result should have been (exact matching):
handshake failed: certificate verify failed due to "certificate has expired"
```

Feels like the tests actually succeed but retrieval of the error message is somehow mangled. It gives the correct failure reason but also says "No error"

**Comment by bohagan1** [comment]
There was an issue with non-error conditions being logged as errors which could overwrite the real error the way TclTLS caches errors. I removed these conditions so hopefully you will only see the real error now. I don't see these errors, but then again didn't see them before the change either. Let me know. Thanks.

**Comment by apnadkarni** [comment]
The spurious messages seem gone. Still a few test failures - see test log below (Windows MinGW64). I suspect many (all?) of these stem from different openssl build options but not sure. You can close this ticket if you think they are benign.

```
TCL_LIBRARY=`cygpath -m C:/src/tcltk/9.0.0/tcl/library` LD_LIBRARY_PATH="/d/src/tcl-installer/sources/TclTLS/build-mingw64:/c/tcl/9.0.0/mingw64/lib:" PATH="/d/src/tcl-installer/sources/TclTLS/build-mingw64:/c/tcl/9.0.0/mingw64/lib:/mingw64/bin:/usr/local/bin:/usr/bin:/bin:/c/Windows/System32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0/:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl" TCLLIBPATH="/d/src/tcl-installer/sources/TclTLS/build-mingw64" /c/tcl/9.0.0/mingw64/bin/tclsh90.exe `cygpath -m ../tests/all.tcl`  \
    -load "package ifneeded tls 1.8.0 \
	[list load `cygpath -m tcl9tls180.dll` [string totitle tls]]"
Tests running in interp:  C:/tcl/9.0.0/mingw64/bin/tclsh90.exe
Tests located in:  D:/src/tcl-installer/sources/TclTLS/tests
Tests running in:  D:/src/tcl-installer/sources/TclTLS/build-mingw64
Temporary files stored in D:/src/tcl-installer/sources/TclTLS/build-mingw64
Test files run in separate interpreters
Running tests that match:  *
Skipping test files that match:  l.*.test
Only running test files that match:  *.test
Tests began at Mon Dec 16 12:31:37 +0530 2024
badssl.test


==== BadSSL-1.10 dh-composite FAILED
==== Contents of test case:

	badssl dh-composite.badssl.com
    
---- Test completed normally; Return code was: 0
---- Return code should have been one of: 1
==== BadSSL-1.10 FAILED



==== BadSSL-1.17 dh1024 FAILED
==== Contents of test case:

	badssl dh1024.badssl.com
    
---- Result was:
handshake failed: unexpected eof while reading
---- Result should have been (exact matching):
handshake failed: dh key too small
==== BadSSL-1.17 FAILED



==== BadSSL-1.27 incomplete-chain FAILED
==== Contents of test case:

	badssl incomplete-chain.badssl.com
    
---- Result was:
couldn't open socket: connection timed out
---- Result should have been (exact matching):
handshake failed: certificate verify failed due to "unable to get local issuer certificate"
==== BadSSL-1.27 FAILED



==== BadSSL-1.28 invalid-expected-sct FAILED
==== Contents of test case:

	badssl invalid-expected-sct.badssl.com
    
---- Result was:
couldn't open socket: connection timed out
---- Result should have been (exact matching):
handshake failed: certificate verify failed due to "unable to get local issuer certificate"
==== BadSSL-1.28 FAILED



==== BadSSL-1.31 mitm-software FAILED
==== Contents of test case:

	badssl mitm-software.badssl.com
    
---- Result was:
couldn't open socket: connection timed out
---- Result should have been (exact matching):
handshake failed: certificate verify failed due to "unable to get local issuer certificate"
==== BadSSL-1.31 FAILED

ciphers.test


==== Ciphers_Protocol_Specific-4.3 TLS1.0 FAILED
==== Contents of test case:

	lcompare [exec_get ":" ciphers -tls1 -s] [::tls::ciphers tls1 0 1]
    
---- Result was:
missing {} unexpected {ECDHE-ECDSA-AES256-SHA ECDHE-RSA-AES256-SHA DHE-RSA-AES256-SHA ECDHE-ECDSA-AES128-SHA ECDHE-RSA-AES128-SHA DHE-RSA-AES128-SHA AES256-SHA AES128-SHA}
---- Result should have been (exact matching):
missing {} unexpected {}
==== Ciphers_Protocol_Specific-4.3 FAILED



==== Ciphers_Protocol_Specific-4.4 TLS1.1 FAILED
==== Contents of test case:

	lcompare [exec_get ":" ciphers -tls1_1 -s] [::tls::ciphers tls1.1 0 1]
    
---- Result was:
missing {} unexpected {ECDHE-ECDSA-AES256-SHA ECDHE-RSA-AES256-SHA DHE-RSA-AES256-SHA ECDHE-ECDSA-AES128-SHA ECDHE-RSA-AES128-SHA DHE-RSA-AES128-SHA AES256-SHA AES128-SHA}
---- Result should have been (exact matching):
missing {} unexpected {}
==== Ciphers_Protocol_Specific-4.4 FAILED



==== Ciphers_Protocol_Specific-4.6 TLS1.3 FAILED
==== Contents of test case:

	lcompare [exec_get ":" ciphers -tls1_3 -s] [::tls::ciphers tls1.3 0 1]
    
---- Result was:
missing {} unexpected {ECDHE-ECDSA-AES256-GCM-SHA384 ECDHE-RSA-AES256-GCM-SHA384 DHE-RSA-AES256-GCM-SHA384 ECDHE-ECDSA-CHACHA20-POLY1305 ECDHE-RSA-CHACHA20-POLY1305 DHE-RSA-CHACHA20-POLY1305 ECDHE-ECDSA-AES128-GCM-SHA256 ECDHE-RSA-AES128-GCM-SHA256 DHE-RSA-AES128-GCM-SHA256 ECDHE-ECDSA-AES256-SHA384 ECDHE-RSA-AES256-SHA384 DHE-RSA-AES256-SHA256 ECDHE-ECDSA-AES128-SHA256 ECDHE-RSA-AES128-SHA256 DHE-RSA-AES128-SHA256 ECDHE-ECDSA-AES256-SHA ECDHE-RSA-AES256-SHA DHE-RSA-AES256-SHA ECDHE-ECDSA-AES128-SHA ECDHE-RSA-AES128-SHA DHE-RSA-AES128-SHA AES256-GCM-SHA384 AES128-GCM-SHA256 AES256-SHA256 AES128-SHA256 AES256-SHA AES128-SHA}
---- Result should have been (exact matching):
missing {} unexpected {}
==== Ciphers_Protocol_Specific-4.6 FAILED

tlsIO.test


==== tlsIO-9.1 testing spurious events FAILED
==== Contents of test case:

    set len 0
    set spurious 0
    set done 0
    proc readlittle {s} {
	global spurious done len
	set l [read $s 1]
	if {[string length $l] == 0} {
	    if {![eof $s]} {
		incr spurious
	    } else {
		close $s
		set done 1
	    }
	} else {
	    incr len [string length $l]
	}
    }
    proc accept {s a p} {
	fconfigure $s -blocking 0
	fileevent $s readable [list do_handshake $s readable readlittle  -buffering none]
    }
    set s [tls::socket  -certfile $serverCert -cafile $caCert -keyfile $serverKey  -server accept 8831]
    set c [tls::socket  -certfile $clientCert -cafile $caCert -keyfile $clientKey  localhost 8831]
    # This differs from socket-9.1 in that both sides need to be
    # non-blocking because of TLS' required handshake
    fconfigure $c -blocking 0
    puts -nonewline $c 01234567890123456789012345678901234567890123456789
    close $c
    set timer [after 10000 "set done timed_out"]
    vwait done
    after cancel $timer
    close $s
    list $spurious $len

---- Result was:
0 0
---- Result should have been (exact matching):
0 50
==== tlsIO-9.1 FAILED



==== tls-bug58-1.0 test protocol negotiation failure FAILED
==== Contents of test case:

    # Following code is based on what was reported in bug #58. Prior
    # to fix the program would crash with a segfault.
    proc Accept {sock args} {
        fconfigure $sock -blocking 0;
        fileevent $sock readable [list Handshake $sock]
    }
    proc Handshake {sock} {
        set ::done HAND
        catch {tls::handshake $sock} msg
        set ::done $msg
    }
    # NOTE: when doing an in-process client/server test, both sides need
    # to be non-blocking for the TLS handshake

    # Server - Only accept TLS 1.2
    set s [tls::socket  -certfile $serverCert -cafile $caCert -keyfile $serverKey -request 0  -require 0 -ssl2 0 -ssl3 0 -tls1 0 -tls1.1 0 -tls1.2 1 -tls1.3 0  -server Accept 8831]
    # Client - Only propose TLS1.0
    set c [tls::socket -async -cafile $caCert -request 0 -require 0  -ssl2 0 -ssl3 0 -tls1 1 -tls1.1 0 -tls1.2 0 -tls1.3 0 localhost 8831]
    fconfigure $c -blocking 0
    puts $c a ; flush $c
    after 5000 [list set ::done timeout]
    vwait ::done
    switch -exact -- $::done {
        "handshake failed: wrong ssl version" -
        "handshake failed: unsupported protocol" {
            set ::done "handshake failed: wrong version number"
        }
    }
    set ::done

---- Test generated error; Return code was: 1
---- Return code should have been one of: 0 2
---- errorInfo: error flushing "sock1e01a052110": software caused connection abort
    while executing
"flush $c"
    ("uplevel" body line 21)
    invoked from within
"uplevel 1 $script"
---- errorCode: POSIX ECONNABORTED {software caused connection abort}
==== tls-bug58-1.0 FAILED


Tests ended at Mon Dec 16 12:37:18 +0530 2024
all.tcl:	Total	155	Passed	107	Skipped	38	Failed	10
Sourced 3 Test Files.
Files with failing tests: badssl.test ciphers.test tlsIO.test
Number of tests skipped for each constraint:
	12	doTestsWithRemoteServer
	1	mac
	12	old_api
	3	ssl2
	3	ssl3
	1	testthread
	2	unexplainedFailure
	4	unixOnly

```

**Comment by apnadkarni** [comment]
I updated to openssl 3.4 (from 3.1) so it would use the Windows native cert store. Test results are below. Out of 7 failures, 2 are connection timeouts, so probably can be ignored, Ciphers_Protocol_Specific-4.3,4.4,4.6 are possibly openssl configuration leaving only 2 - tlsIO-9.1 and tls-bug58-1.0 fail consistently on every run.

```
TCL_LIBRARY=`cygpath -m C:/src/tcltk/9.0.0/tcl/library` LD_LIBRARY_PATH="/d/src/tcl-installer/sources/TclTLS/build-mingw64:/c/tcl/9.0.0/mingw64/lib:" PATH="/d/src/tcl-installer/sources/TclTLS/build-mingw64:/c/tcl/9.0.0/mingw64/lib:/mingw64/bin:/usr/local/bin:/usr/bin:/bin:/c/Windows/System32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0/:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl" TCLLIBPATH="/d/src/tcl-installer/sources/TclTLS/build-mingw64" /c/tcl/9.0.0/mingw64/bin/tclsh90.exe `cygpath -m ../tests/all.tcl` -verbose te \
    -load "package ifneeded tls 1.8.0 \
	[list load `cygpath -m tcl9tls180.dll` [string totitle tls]]"
Tests running in interp:  C:/tcl/9.0.0/mingw64/bin/tclsh90.exe
Tests located in:  D:/src/tcl-installer/sources/TclTLS/tests
Tests running in:  D:/src/tcl-installer/sources/TclTLS/build-mingw64
Temporary files stored in D:/src/tcl-installer/sources/TclTLS/build-mingw64
Test files run in separate interpreters
Running tests that match:  *
Skipping test files that match:  l.*.test
Only running test files that match:  *.test
Tests began at Mon Dec 16 13:30:19 +0530 2024
badssl.test
---- BadSSL-1.1 start
---- BadSSL-1.2 start
---- BadSSL-1.3 start
---- BadSSL-1.5 start
---- BadSSL-1.6 start
---- BadSSL-1.7 start
---- BadSSL-1.8 start
---- BadSSL-1.10 start
---- BadSSL-1.11 start
---- BadSSL-1.13 start
---- BadSSL-1.17 start
---- BadSSL-1.18 start
---- BadSSL-1.19 start
---- BadSSL-1.20 start
---- BadSSL-1.21 start
---- BadSSL-1.22 start
---- BadSSL-1.23 start
---- BadSSL-1.24 start
---- BadSSL-1.25 start
---- BadSSL-1.26 start
---- BadSSL-1.27 start
---- BadSSL-1.28 start
---- BadSSL-1.29 start
---- BadSSL-1.30 start
---- BadSSL-1.31 start
---- BadSSL-1.32 start
---- BadSSL-1.33 start


==== BadSSL-1.33 no-sct FAILED
---- Test generated error; Return code was: 1
---- Return code should have been one of: 0 2
---- errorInfo: couldn't open socket: connection timed out
    while executing
"::socket no-sct.badssl.com 443"
    ("eval" body line 1)
    invoked from within
"eval $socketCmd $sopts"
    (procedure "tls::socket" line 70)
    invoked from within
"tls::socket -autoservername 1 -require 1 no-sct.badssl.com 443"
    ("eval" body line 1)
    invoked from within
"eval $cmd"
    (procedure "badssl" line 1)
    invoked from within
"badssl no-sct.badssl.com"
    ("uplevel" body line 2)
    invoked from within
"uplevel 1 $script"
---- errorCode: POSIX ETIMEDOUT {connection timed out}
==== BadSSL-1.33 FAILED

---- BadSSL-1.34 start
---- BadSSL-1.35 start
---- BadSSL-1.36 start
---- BadSSL-1.37 start
---- BadSSL-1.38 start
---- BadSSL-1.39 start
---- BadSSL-1.40 start
---- BadSSL-1.41 start
---- BadSSL-1.42 start
---- BadSSL-1.43 start
---- BadSSL-1.44 start
---- BadSSL-1.46 start
---- BadSSL-1.47 start
---- BadSSL-1.49 start
---- BadSSL-1.50 start
---- BadSSL-1.51 start
---- BadSSL-1.52 start
---- BadSSL-1.53 start


==== BadSSL-1.53 sha512 FAILED
---- Result was:
couldn't open socket: connection timed out
---- Result should have been (exact matching):
handshake failed: certificate verify failed due to "certificate has expired"
==== BadSSL-1.53 FAILED

---- BadSSL-1.54 start
---- BadSSL-1.56 start
---- BadSSL-1.57 start
---- BadSSL-1.59 start
---- BadSSL-1.61 start
---- BadSSL-1.62 start
---- BadSSL-1.64 start
---- BadSSL-1.65 start
---- BadSSL-1.66 start
---- BadSSL-1.68 start
---- BadSSL-1.69 start
ciphers.test
---- Protocols-1.1 start
---- Ciphers_By_Protocol-2.3 start
---- Ciphers_By_Protocol-2.4 start
---- Ciphers_By_Protocol-2.5 start
---- Ciphers_By_Protocol-2.6 start
---- Ciphers_With_Descriptions-3.3 start
---- Ciphers_With_Descriptions-3.4 start
---- Ciphers_With_Descriptions-3.5 start
---- Ciphers_With_Descriptions-3.6 start
---- Ciphers_Protocol_Specific-4.3 start


==== Ciphers_Protocol_Specific-4.3 TLS1.0 FAILED
---- Result was:
missing {} unexpected {ECDHE-ECDSA-AES256-SHA ECDHE-RSA-AES256-SHA DHE-RSA-AES256-SHA ECDHE-ECDSA-AES128-SHA ECDHE-RSA-AES128-SHA DHE-RSA-AES128-SHA AES256-SHA AES128-SHA}
---- Result should have been (exact matching):
missing {} unexpected {}
==== Ciphers_Protocol_Specific-4.3 FAILED

---- Ciphers_Protocol_Specific-4.4 start


==== Ciphers_Protocol_Specific-4.4 TLS1.1 FAILED
---- Result was:
missing {} unexpected {ECDHE-ECDSA-AES256-SHA ECDHE-RSA-AES256-SHA DHE-RSA-AES256-SHA ECDHE-ECDSA-AES128-SHA ECDHE-RSA-AES128-SHA DHE-RSA-AES128-SHA AES256-SHA AES128-SHA}
---- Result should have been (exact matching):
missing {} unexpected {}
==== Ciphers_Protocol_Specific-4.4 FAILED

---- Ciphers_Protocol_Specific-4.5 start
---- Ciphers_Protocol_Specific-4.6 start


==== Ciphers_Protocol_Specific-4.6 TLS1.3 FAILED
---- Result was:
missing {} unexpected {ECDHE-ECDSA-AES256-GCM-SHA384 ECDHE-RSA-AES256-GCM-SHA384 DHE-RSA-AES256-GCM-SHA384 ECDHE-ECDSA-CHACHA20-POLY1305 ECDHE-RSA-CHACHA20-POLY1305 DHE-RSA-CHACHA20-POLY1305 ECDHE-ECDSA-AES128-GCM-SHA256 ECDHE-RSA-AES128-GCM-SHA256 DHE-RSA-AES128-GCM-SHA256 ECDHE-ECDSA-AES256-SHA384 ECDHE-RSA-AES256-SHA384 DHE-RSA-AES256-SHA256 ECDHE-ECDSA-AES128-SHA256 ECDHE-RSA-AES128-SHA256 DHE-RSA-AES128-SHA256 ECDHE-ECDSA-AES256-SHA ECDHE-RSA-AES256-SHA DHE-RSA-AES256-SHA ECDHE-ECDSA-AES128-SHA ECDHE-RSA-AES128-SHA DHE-RSA-AES128-SHA AES256-GCM-SHA384 AES128-GCM-SHA256 AES256-SHA256 AES128-SHA256 AES256-SHA AES128-SHA}
---- Result should have been (exact matching):
missing {} unexpected {}
==== Ciphers_Protocol_Specific-4.6 FAILED

---- Version-5.1 start
---- Version-5.2 start
tlsIO.test
Skipping tests with remote server. See tests/socket.test for
information on how to run remote server.
Reason for not doing remote tests: couldn't open socket: connection refused
---- tlsIO-1.1 start
---- tlsIO-1.2 start
---- tlsIO-1.3 start
---- tlsIO-1.4 start
---- tlsIO-1.5 start
---- tlsIO-1.6 start
---- tlsIO-1.7 start
---- tlsIO-1.8 start
---- tlsIO-1.9 start
---- tlsIO-1.10 start
---- tlsIO-1.11 start
---- tlsIO-1.12 start
---- tlsIO-2.1 start
---- tlsIO-2.2 start
---- tlsIO-2.3 start
---- tlsIO-2.4 start
---- tlsIO-2.5 start
---- tlsIO-2.6 start
---- tlsIO-2.7 start
---- tlsIO-2.8 start
---- tlsIO-2.9 start
---- tlsIO-2.10 start
---- tlsIO-2.11 start
---- tlsIO-3.1 start
---- tlsIO-3.2 start
---- tlsIO-4.1 start
---- tlsIO-4.2 start
---- tlsIO-5.2 start
---- tlsIO-6.1 start
---- tlsIO-7.1 start
---- tlsIO-7.2 start
---- tlsIO-7.3 start
---- tlsIO-7.4 start
---- tlsIO-7.5 start
---- tlsIO-8.1 start
---- tlsIO-9.1 start


==== tlsIO-9.1 testing spurious events FAILED
---- Result was:
0 0
---- Result should have been (exact matching):
0 50
==== tlsIO-9.1 FAILED

---- tlsIO-9.2 start
---- tlsIO-10.1 start
---- tlsIO-12.1 start
---- tlsIO-12.2 start
---- tlsIO-14.1 start
---- tlsIO-14.2 start
---- tlsIO-14.3 start
---- tlsIO-14.4 start
---- tlsIO-14.5 start
---- tls-bug58-1.0 start


==== tls-bug58-1.0 test protocol negotiation failure FAILED
---- Test generated error; Return code was: 1
---- Return code should have been one of: 0 2
---- errorInfo: error flushing "sock1ea34f38b40": software caused connection abort
    while executing
"flush $c"
    ("uplevel" body line 21)
    invoked from within
"uplevel 1 $script"
---- errorCode: POSIX ECONNABORTED {software caused connection abort}
==== tls-bug58-1.0 FAILED


Tests ended at Mon Dec 16 13:34:31 +0530 2024
all.tcl:	Total	155	Passed	110	Skipped	38	Failed	7
Sourced 3 Test Files.
Files with failing tests: badssl.test ciphers.test tlsIO.test
Number of tests skipped for each constraint:
	12	doTestsWithRemoteServer
	1	mac
	12	old_api
	3	ssl2
	3	ssl3
	1	testthread
	2	unexplainedFailure
	4	unixOnly
```

**Comment by bohagan1** [comment]
The cipher tests will fail until I update them to use the new OpenSSL 3.x APIs. That’s not for this release. So known issue. I fixed tls-bug58-1.0 yesterday, so check the repo. I also almost fixed tlsIO-9.1. There is one issue I can’t track down causing a spurious empty read. I don’t see any failures in the BadSSL test cases, so I’ll need to see if that’s a Windows version or OpenSSL version issue.

**Comment by apnadkarni** [comment]
Closing as not seeing the original reported messages now.
